sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaHeader.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.kafka;
+
+/**
+ * This is a copy of Kafka's {@link org.apache.kafka.common.header.Header}. Included here in order
+ * to support older Kafka versions (0.9.x).
+ */
+public interface KafkaHeader {
```

Comment:
I think strictly we don't need to our own versions `Header`, `Headers` etc. How about this: 
 - Add 'Headers' field to KafkaRecord. We ensure the field is accessed only if the `consumerSpEL.hasHeaders` is true.
 - add `Headers getHeaders()` method to KafkaRecord, which throws an exception if `hasHeaders` is false.

I think we could have done something like this with `TimestampType` too. WDYT.
It is a bit hacky, but I guess it is better than replicating more of ConsumerRecord functionality.

Reply:
> Add 'Headers' field to KafkaRecord

Is 'Headers' referring to ```org.apache.kafka.common.header.Headers```? If so, how would this approach work for Kafka 0.9.x (note that this interface is only available in Kafka 0.11.x) ? Wouldn't this lead to a compile-time error?

The main reason we use our own versions of ```Header```, etc is to avoid such compile-time errors.

Reply:
We want to avoid runtime errors, not compile time. As long as user with older version of Kafka at runtime does not use getHeaders() interface, it will be fine. I tried it in a test program and it worked ok.

Reply:
We can require recent kafka version to compile and run tests. But at runtime we want KafkaIO to work with older versions of kafka-clients at runtime. That's why we have the runtime reflection checks.

Reply:
Hmm, when adding ```Headers``` to ```KafkaRecord```, I was not able to instantiate ```KafkaRecord``` in my test program. Compilation failed with:

```
Error:(8, 5) java: cannot access org.apache.kafka.common.header.Headers
  class file for org.apache.kafka.common.header.Headers not found
```
Tried with both JDK 1.7 and 1.8. This is using kafka-clients 0.9.0.0.

Reply:
As noted in previous two comments, we don't need to support compiling with 0.9. Even master would not not compile with 0.9. We can require 0.11.x (whichever version first included headers) for compilation and to run tests.
If you get the code working with the default version of kafka-clients Beam (I think 1.0.x), that is good enough. During the review we can check if it would work with older version of kafka-clients at runtime.

Reply:
Yes, I understand. I have removed the additional classes. What I mean to say is if the test program (which depends on beam-kafka) calls KafkaRecord.getHeaders(), this raises a compile error. Is this what we want? 

It seems a bit counter-intuitive to expose a method such as getHeaders() and not be able to compile the program when using it.

Reply:
I see. That is correct, compilation error is fine. All the kafka-client versions before 0.10.1 are already deprecated [1]. If a test wants to test kafka headers, it probably makes sense to depend on a kafka-client version that has headers. 

[1]https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L624